anxcloud/data_source_vlan.go

```diff
@@ -0,0 +1,111 @@
+package anxcloud
+
+import (
+	"context"
+	"net/url"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"go.anx.io/go-anxcloud/pkg/vlan"
+)
+
+func dataSourceVLAN() *schema.Resource {
+	return &schema.Resource{
+		Description: "Provides details about an Anexia Cloud VLAN. This data source is useful if you want to use a non-terraform managed VLAN.",
+		ReadContext: dataSourceVLANRead,
+
+		// TODO: can I reuse the existing schema?
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				Description:  identifierDescription,
+				ExactlyOneOf: []string{"id", "name"},
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "The name of the VLAN.",
+			},
+			"location_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "ANX Location Identifier.",
+			},
+			"vm_provisioning": {
+				Type:        schema.TypeBool,
+				Computed:    true,
+				Description: "True if VM provisioning shall be enabled. Defaults to false.",
+			},
+			"description_customer": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Additional customer description.",
+			},
+			"description_internal": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Internal description.",
+			},
+			"role_text": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "Role of the VLAN.",
+			},
+			"status": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "VLAN status.",
+			},
+			"locations": schemaLocations(),
+		},
+	}
+}
+
+func dataSourceVLANRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(providerContext).legacyClient
+	v := vlan.NewAPI(c)
+
+	var foundVLANID string
+
+	// First, we search for the VLAN via the name.
+	if nameFromSchema, ok := d.GetOk("name"); ok {
+		name := nameFromSchema.(string)
+
+		vlans, err := listAllPages(func(page int) ([]vlan.Summary, error) {
+			return v.List(ctx, page, 100, url.QueryEscape(name))
+		})
+		if err != nil {
+			return diag.FromErr(err)
```

Comment:
Please wrap `err` with a suitable message. Engine messages can sometimes be quite vague.